### PR TITLE
Add separate option for verifying incoming HTTPS traffic

### LINKED ISSUE #2969: Make it possible turn off verify incoming TLS for HTTPS endpoints
Many users want verification in both directions for RPC, but want to let normally-configured web browsers hit the https endpoints with TLS but no client certificates. It should be possible to configure this (and it should be opt-in, since we want to insist on client certs for HTTPS unless they specifically don't want this).

### PR DESCRIPTION
**Pull request by kyhavlov**
Resolves #2969

### CODE REVIEW

**Comment by slackpad** [comment]
To keep backward compatibility with the existing sense of `verify_incoming` I'm thinking we should add another RPC option and do logic like this:

```
actuallyVerifyIncomingHTTPS := verify_incoming || verify_incoming_https
actuallyVerifyIncomingRPC := verify_incoming || verify_incoming_rpc
```

File: testutil/server.go [hunk 1]
```diff
@@ -56,32 +56,33 @@ type TestAddressConfig struct {
...
+	ACLMasterToken      string                 `json:"acl_master_token,omitempty"`
+	ACLDatacenter       string                 `json:"acl_datacenter,omitempty"`
+	ACLDefaultPolicy    string                 `json:"acl_default_policy,omitempty"`
+	ACLEnforceVersion8  bool                   `json:"acl_enforce_version_8"`
+	Encrypt             string                 `json:"encrypt,omitempty"`
+	CAFile              string                 `json:"ca_file,omitempty"`
+	CertFile            string                 `json:"cert_file,omitempty"`
+	KeyFile             string                 `json:"key_file,omitempty"`
+	VerifyIncoming      bool                   `json:"verify_incoming,omitempty"`
+	VerifyIncomingHTTPS bool                   `json:"verify_incoming_https,omitempty"`
```
Comment (magiconair) [inline testutil/server.go]: Don't you need a `VerifyIncomingRPC` as well here?

File: website/source/docs/agent/options.html.markdown [hunk 2]
```diff
@@ -1052,9 +1052,21 @@ Consul will not enable TLS for the HTTP API unless the `https` port has been ass
...
   set to true, Consul requires that all incoming
   connections make use of TLS and that the client provides a certificate signed
+  by the Certificate Authority from the [`ca_file`](#ca_file). This applies to both
+  server RPC and to the HTTPS API. By default, this is false, and
+  Consul will not enforce the use of TLS or verify a client's authenticity.
+
+* <a name="verify_incoming_rpc"></a><a href="#verify_incoming_rpc">`verify_incoming_rpc`</a> - If
+  set to true, Consul requires that all incoming RPC
+  connections make use of TLS and that the client provides a certificate signed
+  by the Certificate Authority from the [`ca_file`](#ca_file). By default, this is false, and
```
Comment (magiconair) [inline website/source/docs/agent/options.html.markdown]: Doesn't that need updating with the recent CAPath change you were working on?

File: website/source/docs/agent/options.html.markdown [hunk 3]
```diff
@@ -1052,9 +1052,21 @@ Consul will not enable TLS for the HTTP API unless the `https` port has been ass
...
+  Consul will not enforce the use of TLS or verify a client's authenticity.
+
+* <a name="verify_incoming_https"></a><a href="#verify_incoming_https">`verify_incoming_https`</a> - If
+  set to true, Consul requires that all incoming HTTPS
+  connections make use of TLS and that the client provides a certificate signed
   by the Certificate Authority from the [`ca_file`](#ca_file). By default, this is false, and
```
Comment (magiconair) [inline website/source/docs/agent/options.html.markdown]: same here

File: testutil/server.go (hunk 1 above)
Comment (kyhavlov) [inline testutil/server.go]: yeah, it should be here for completeness

File: website/source/docs/agent/options.html.markdown (hunk 2 above)
Comment (kyhavlov) [inline website/source/docs/agent/options.html.markdown]: Yes, good catch

File: api/api_test.go [hunk 4]
```diff
@@ -264,45 +265,102 @@ func TestClientTLSOptions(t *testing.T) {
...
-	if err != nil {
-		t.Fatal(err)
-	}
 
-	// Should succeed
-	_, err = clientWithCerts.Agent().Self()
-	if err != nil {
-		t.Fatal(err)
-	}
+	// Start a server without VerifyIncomingHTTPS
```
Comment (magiconair) [inline api/api_test.go]: Move this to the top and give it a different name, e.g. `srvNoVerify`

File: api/api_test.go [hunk 5]
```diff
@@ -256,6 +256,7 @@ func TestSetupTLSConfig(t *testing.T) {
 
 func TestClientTLSOptions(t *testing.T) {
 	t.Parallel()
+	// Start a server that verifies incoming HTTPS connections
 	_, s := makeClientWithConfig(t, nil, func(conf *testutil.TestServerConfig) {
```
Comment (magiconair) [inline api/api_test.go]: s/s/srvVerify/
Don't you have to set `conf.VerifyIncomingHTTPS = true` ?

Comment (kyhavlov) [inline api/api_test.go]: it is set to true, right below this

**Comment by magiconair** [comment]
LGTM

**Comment by magiconair** [comment]
Please squash the change

**Pull request merged**